WordPress/src/main/java/org/wordpress/android/ui/main/SitePickerAdapter.java

```diff
@@ -541,16 +541,22 @@ public int compare(SiteRecord site1, SiteRecord site2) {
             }
 
             if (mSites == null || !mSites.isSameList(sites)) {
-                mAllSites = (SiteList) sites.clone();
-                mSites = filteredSitesByTextIfInSearchMode(sites);
+                SiteList allSites = (SiteList) sites.clone();
+                SiteList filteredSites = filteredSitesByTextIfInSearchMode(sites);
+
+                return new SiteList[]{allSites, filteredSites};
             }
 
             return null;
         }
 
         @Override
-        protected void onPostExecute(Void results) {
-            notifyDataSetChanged();
+        protected void onPostExecute(SiteList[] updatedSiteLists) {
+            if (updatedSiteLists != null) {
```

Comment:
> I might still be missing some case so, can you describe a scenario in this particular Adapter/Fragment/Activity combo that the synchronized would help avoid an error?

The kind of things I can think of atm are better suited for being discussed with beers at this point.

>  it's better to have a specific theory to test and in this case,  the synchronized block is not testing/guarding any specific theory/case.

Theory:
> The fix tries to put the instance data (mSites) change closer to the notifyDataSetChanged() call so the adapter has consistent data as soon as possible.

Basically, the way it looks to me is this current PR is hoping for things to happen, I'm suggesting there is a way to guarantee this will happen.